org.adempiere.base/src/org/compiere/model/MInvoice.java

```diff
@@ -2319,8 +2321,8 @@ else if (getGrandTotal().compareTo(totalPayAmt) != 0 && ptList.size() > 0)
 				}
 			}
 		}
-
-		if (PAYMENTRULE_Cash.equals(getPaymentRule())) {
+		
+		if (PAYMENTRULE_Cash.equals(getPaymentRule()) || allocateFromOrder()) {

```

Comment:
why is this required now and was not required before with POS Orders?

I have not tested but I feel there can be some collaterals here.

is there a test case that validates that old orders/invoices keep running fine?

Reply:
Important to note that - allocateFromOrder get payment with c_order_id, in other word, should only works for prepay order. The flow for POS order is different, payment is auto created from invoice.

Also, we do have existing test case that should cover some of the order, shipment, invoice and payment flow.

Reply:
So If I understand well, its fine to leave allocateFromOrder() as it is now, if it won't affect the POS Orders.

Reply:
I have not tested this yet, but the thing is that the new method allocateFromOrder is just needed if the invoice was not generated on complete of the prepaid order (the new use case covered by this ticket).

When is not an prepaid order, or when the invoice was generated on complete, everything is working fine without this new method.

So, what I understand that @hengsin advices is to call allocateFromOrder just for this new case (prepaid order that didn't generate invoice immediately).